nmclicktests/environment.js

```diff
@@ -314,38 +88,225 @@ class Environment {
+  }
+
+  async shutdown() {
+    this.shuttinDown = true;
+
+    await this.backgroundAction('POST', '/api/testing/cleanup')
+
+    if (this.ungitServerProcess) {
+      this.ungitServerProcess.kill('SIGINT');
+      this.ungitServerProcess = null;
+    }
+
+    if (this.browser) {
+      await this.browser.close();
+      this.browser = null;
+      this.page = null;
+    }
+  }
+
+  // server helpers
+
+  async backgroundAction(method, url, body) {
+    url = this.getRootUrl() + url;
+
+    let req;
+    if (method === 'GET') {
+      req = request.get(url).withCredentials().query(body);
+    } else if (method === 'POST') {
+      req = request.post(url).send(body);
+    } else if (method === 'DELETE') {
+      req = request.delete(url).send(body);
+    }
+
+    req.set({ 'encoding': 'utf8', 'cache-control': 'no-cache', 'Content-Type': 'application/json' });
+
+    const response = await req;
+    return response.body;
+  }
+
+  async createRepos(testRepoPaths, config) {
+    for (let i = 0; i < config.length; i++) {
+      const conf = config[i];
+      conf.bare = !!conf.bare;
+      await this.initRepo(conf);
+      await this.createCommits(conf, conf.initCommits);
+      testRepoPaths.push(conf.path);
+    }
+  }
+
+  async initRepo(options) {
+    if (options.path) {
+      await rimraf(options.path);
+      await mkdirp(options.path);
+    } else {
+      winston.info('Creating temp folder');
+      options.path = await this.createTempFolder();
+    }
+    await this.backgroundAction('POST', '/api/init', options);
+  }
+
+  async createTempFolder() {
+    const res = await this.backgroundAction('POST', '/api/testing/createtempdir');
+    return res.path;
+  }
+
+  async createCommits(config, limit, x) {
+    x = x || 0
+    if (!limit || limit < 0 || x === limit) return;
+
+    await this.createTestFile(`${config.path}/testy${x}`);
+    await this.backgroundAction('POST', '/api/commit', {
+      path: config.path,
+      message: `Init Commit ${x}`,
+      files: [{ name: `testy${x}` }]
     });
-    ungitServer.on('exit', () => winston.info('UNGIT SERVER EXITED'));
+    await this.createCommits(config, limit, x + 1);
+  }
+
+  createTestFile(filename, repoPath) {
+    return this.backgroundAction('POST', '/api/testing/createfile', { file: filename, path: repoPath });
+  }
 
-    this.ungitServerProcess = ungitServer;
+  // browser helpers
+
+  async goto(url, options) {
+    if (!this.page) {
+      const pages = await this.browser.pages();
+      const page = this.page = pages[0];
+
+      page.on('console', (message) => {
+        const text = `[ui ${message.type()}] ${(new Date()).toISOString()}  - ${message.text()} ${message.args().join(', ')}`;
+
+        if (message.type() === 'error' && !this.shuttinDown) {
+          winston.error(text);
+        } else {
+          winston.info(text);
+        }
+      });
+    }
 
-    return Bluebird.resolve();
+    await this.page.goto(url);
+
+    if (!options || options.waitForSocketIO) {
```

Comment:
Puppeteer is so fast that we have to manually wait for socket io to connect, otherwise this can lead to problems when code relies on working directory changes.